applications/tari_validator_node/src/p2p/services/mempool/service.rs

```diff
@@ -586,21 +586,14 @@ where
         };
 
         let current_epoch = self.epoch_manager.current_epoch().await?;
-        let local_committee_shard = self.epoch_manager.get_local_committee_shard(current_epoch).await?;
-        let is_input_shard = local_committee_shard.includes_any_shard(executed.transaction().all_inputs_iter());
-
-        if should_propagate && is_input_shard {
-            // Forward the transaction to any output shards that are not part of the input shard set as these have
-            // already been forwarded
-            let num_committees = self.epoch_manager.get_num_committees(current_epoch).await?;
-            let input_buckets = executed
-                .involved_shards_iter()
-                .map(|s| s.to_committee_bucket(num_committees))
-                .collect::<HashSet<_>>();
+
+        // TODO: Should we always propagate? The TX is initially send only to single output shard (TxReceipt).
+        if should_propagate {
```

Comment:
Please explain this further. The rule is that only input shards can send to output shards, happy to change that rule but there should be a reason. If there is a problem(bug) with the implementation then that should be fixed.

Reply:
We propagate to all input shards before execution. This propagation happens after execution, and that should only be sent to output shards since input shard propagation has already happened.

Reply:
The TxReceipt is a special case, so we could/should propagate before execution to the txreceipt shard

Reply:
Added back the filtering. Working on the distribution via TxReceipt only.

Reply:
I think we can readd the is_input_shard check now (since it will result in messages being sent by output shards to the output shards again). And have the input shards propagate before execution to the TxReciept output shard. We already handle the [special case when receiving a transaction for output shards](https://github.com/tari-project/tari-dan/blob/9c2f369f69c27b570fce96b5316e1f62e694a032/applications/tari_validator_node/src/p2p/services/mempool/service.rs#L254) and the transaction has no inputs.

Reply:
So I think the only change that is needed is to include the tx receipt in propagation [here](https://github.com/tari-project/tari-dan/blob/9c2f369f69c27b570fce96b5316e1f62e694a032/applications/tari_validator_node/src/p2p/services/mempool/service.rs#L356)

```rust
   transaction.involved_shards_iter().copied().chain(Some(tx_shard_id)).collect(),
```

Reply:
Ah ok, so we would have other output shards too if the account is created from free coins, not just the TxReciept.

In this case, we also need to propagate to non-TxReceipt output shards after execution. So maybe we should add another cluase to the if for this case rather than always propagating? So that "normal" transactions don't result in output shards repropagating to other output shards needlessly.

```rust
       if should_propagate && (is_input_shard || executed.transaction().num_involved_shards() == 0) {
```

Reply:
What I do now is that indexer sends the TX only to transaction id shard. That propagates to known (including filled) input shards. After the execution each if we are the transaction id shard we propagate it to any shard that was not in the inputs (i.e. output shards). So there is much less propagation going around, only from the transaction id shard. And only once to every other bucket.